src/Xamarin.Android.Build.Tasks/Tasks/Aapt2.cs

```diff
@@ -0,0 +1,164 @@
+using System.Xml.Linq;
+using Microsoft.Build.Utilities;
+using Microsoft.Build.Framework;
+using System.Text.RegularExpressions;
+using System.Collections.Generic;
+using Xamarin.Android.Tools;
+using ThreadingTasks = System.Threading.Tasks;
+
+namespace Xamarin.Android.Tasks {
+	
+	public class Aapt2 : AsyncTask {
+
+		protected Dictionary<string, string> resource_name_case_map = new Dictionary<string, string> ();
+
+		[Required]
+		public string ResourceDirectory { get; set; }
+
+		public string ResourceNameCaseMap { get; set; }
+
+		public string ResourceSymbolsTextFile { get; set; }
+
+		protected string ToolName { get { return OS.IsWindows ? "aapt2.exe" : "aapt2"; } }
+
+		public string ToolPath { get; set; }
+
+		public string ToolExe { get; set; }
+
+		protected string ResourceDirectoryFullPath {
+			get {
+				return (Path.IsPathRooted (ResourceDirectory) ? ResourceDirectory : Path.Combine (WorkingDirectory, ResourceDirectory)).TrimEnd ('\\');
+			}
+		}
+
+		protected string GenerateFullPathToTool ()
+		{
+			return Path.Combine (ToolPath, string.IsNullOrEmpty (ToolExe) ? ToolName : ToolExe);
+		}
+
+		protected bool RunAapt (string commandLine, IList<OutputLine> output)
+		{
+			var stdout_completed = new ManualResetEvent (false);
+			var stderr_completed = new ManualResetEvent (false);
+			var psi = new ProcessStartInfo () {
+				FileName = GenerateFullPathToTool (),
+				Arguments = commandLine,
+				UseShellExecute = false,
+				RedirectStandardOutput = true,
+				RedirectStandardError = true,
+				CreateNoWindow = true,
+				WindowStyle = ProcessWindowStyle.Hidden,
+				WorkingDirectory = WorkingDirectory,
+			};
+			object lockObject = new object ();
+			using (var proc = new Process ()) {
+				proc.OutputDataReceived += (sender, e) => {
+					if (e.Data != null)
+						lock (lockObject)
+							output.Add (new OutputLine (e.Data, stdError: false));
+					else
+						stdout_completed.Set ();
+				};
+				proc.ErrorDataReceived += (sender, e) => {
+					if (e.Data != null)
+						lock (lockObject)
+							output.Add (new OutputLine (e.Data, stdError: true));
+					else
+						stderr_completed.Set ();
+				};
+				LogDebugMessage ("Executing {0}", commandLine);
+				proc.StartInfo = psi;
+				proc.Start ();
+				proc.BeginOutputReadLine ();
+				proc.BeginErrorReadLine ();
+				Token.Register (() => {
+					try {
+						proc.Kill ();
+					} catch (Exception) {
+					}
+				});
+				proc.WaitForExit ();
+				if (psi.RedirectStandardError)
+					stderr_completed.WaitOne (TimeSpan.FromSeconds (30));
+				if (psi.RedirectStandardOutput)
+					stdout_completed.WaitOne (TimeSpan.FromSeconds (30));
+				return proc.ExitCode == 0 && !output.Any (x => x.StdError);
+			}
+		}
+
+		protected void LogEventsFromTextOutput (string singleLine, MessageImportance messageImportance, bool apptResult)
+		{
+			if (string.IsNullOrEmpty (singleLine))
+				return;
+
+			var match = AndroidToolTask.AndroidErrorRegex.Match (singleLine.Trim ());
```

Comment:
Reading this and the following code makes me think that at some point -- not necessarily *now* -- we should better wrap this up, so instead of string-ized `match.Groups ["file"]`, we can use an actual `error.FileName` property...